README.md
# [DEPRECATED] hapi-magic-filter

This package is deprecated and not actively developed anymore. Read the [deprecation notice](https://github.com/ruiquelhas/hapi-magic-filter/issues/16) for more details on how to move forward.


CHANGELOG.md
# Changelog

## [v2.0.1](https://github.com/ruiquelhas/hapi-magic-filter/tree/v2.0.1) (2016-06-15)
[Full Changelog](https://github.com/ruiquelhas/hapi-magic-filter/compare/v2.0.0...v2.0.1)

**Merged pull requests:**

- Deprecate package [\#19](https://github.com/ruiquelhas/hapi-magic-filter/pull/19) ([ruiquelhas](https://github.com/ruiquelhas))

## [v2.0.0](https://github.com/ruiquelhas/hapi-magic-filter/tree/v2.0.0) (2015-11-01)
[Full Changelog](https://github.com/ruiquelhas/hapi-magic-filter/compare/v1.1.0...v2.0.0)

**Merged pull requests:**

- Update wreck to version 7.0.0 🚀 [\#9](https://github.com/ruiquelhas/hapi-magic-filter/pull/9) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))
- Update boom to version 3.0.0 🚀 [\#8](https://github.com/ruiquelhas/hapi-magic-filter/pull/8) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))
- Update lab to version 7.1.0 🚀 [\#7](https://github.com/ruiquelhas/hapi-magic-filter/pull/7) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))
- Update code to version 2.0.1 🚀 [\#6](https://github.com/ruiquelhas/hapi-magic-filter/pull/6) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))
- Update code to version 2.0.0 🚀 [\#5](https://github.com/ruiquelhas/hapi-magic-filter/pull/5) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))
- Update lab to version 7.0.0 🚀 [\#4](https://github.com/ruiquelhas/hapi-magic-filter/pull/4) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))

## [v1.1.0](https://github.com/ruiquelhas/hapi-magic-filter/tree/v1.1.0) (2015-10-21)
[Full Changelog](https://github.com/ruiquelhas/hapi-magic-filter/compare/v1.0.1...v1.1.0)

**Merged pull requests:**

- Update all dependencies 🌴 [\#3](https://github.com/ruiquelhas/hapi-magic-filter/pull/3) ([greenkeeperio-bot](https://github.com/greenkeeperio-bot))

## [v1.0.1](https://github.com/ruiquelhas/hapi-magic-filter/tree/v1.0.1) (2015-02-07)
[Full Changelog](https://github.com/ruiquelhas/hapi-magic-filter/compare/v1.0.0...v1.0.1)

## [v1.0.0](https://github.com/ruiquelhas/hapi-magic-filter/tree/v1.0.0) (2015-01-07)
[Full Changelog](https://github.com/ruiquelhas/hapi-magic-filter/compare/v0.1.0...v1.0.0)

**Fixed bugs:**

- Malformed JSON response  [\#2](https://github.com/ruiquelhas/hapi-magic-filter/issues/2)
- stream.pause\(\) not compatible with node v0.10? [\#1](https://github.com/ruiquelhas/hapi-magic-filter/issues/1)

## [v0.1.0](https://github.com/ruiquelhas/hapi-magic-filter/tree/v0.1.0) (2014-08-09)


test/edge.js
var fs = require('fs');

var Lab = require('lab');
var Hapi = require('hapi');
var streamToPromise = require('stream-to-promise');
var Form = require('form-data');

var plugin = require('../lib/index');

Lab.experiment('general plugin edge-cases', function () {
  var suite = {};

  Lab.before(function (done) {
    suite.server = Hapi.createServer();
    suite.server.pack.register({
      plugin: plugin,
      options: {
        allowed: {
          'png': '8950'
        }
      }
    }, function (err) {
      suite.server.route({
        path: '/',
        method: ['GET', 'POST'],
        handler: function (request, reply) {
          return reply();
        }
      });

      done();
    });
  });

  Lab.test('returns control to the app if Content-Type is not multipart/form-data', function (done) {
    suite.request = { url: '/', method: 'POST' };
    suite.server.inject(suite.request, function (response) {
      Lab.expect(response.statusCode).to.equal(200);
      done();
    });
  });

  Lab.test('returns control to the app on GET requests', function (done) {
    suite.request = { url: '/', method: 'GET' };
    suite.server.inject(suite.request, function (response) {
      Lab.expect(response.statusCode).to.equal(200);
      done();
    });
  });

  Lab.test('returns error if some file is not supported', function (done) {
    var form = new Form();
    form.append('text', 'dummy');
    form.append('file', fs.createReadStream('resources/file.gif'));

    streamToPromise(form).then(function (payload) {
      suite.request = {
        headers: form.getHeaders(),
        method: 'POST',
        payload: payload,
        url: '/'
      };

      suite.server.inject(suite.request, function (response) {
        Lab.expect(response.statusCode).to.equal(415);
        done();
      });
    });
  });
});


test/index.js
'use strict';

var fs = require('fs');

var Lab = require('lab');
var lab = exports.lab = Lab.script();

var common = require('./common');

lab.describe('hapi-magic-filter', function () {

    var server, payload;

    lab.afterEach(function (done) {

        fs.unlink('test/tmp/file', function () {
            return done();
        });
    });

    lab.describe('Data parsing-enabled file upload', function () {

        lab.before(function (done) {

            payload = {
                output: 'data',
                parse: true
            };

            var handler = function (request, reply) {
                var buffer = request.payload.file;
                var tmp = 'test/tmp/file';

                var done = function (err) {
                    if (err) {
                        return reply(err);
                    }

                    return reply();
                };

                var save = function (err, fd) {
                    if (err) {
                        return reply(err);
                    }

                    fs.write(fd, buffer, 0, buffer.length, 0, done);
                };

                fs.open(tmp, 'w+', save);
            };

            var config = {
                payload: payload,
                handler: handler
            };

            common.boostrap(config, function (err, instance) {

                if (err) {
                    return done(err);
                }

                server = instance;
                return done();
            });
        });

        lab.test('Returns OK if media type is supported.', function (done) {
            common.verifyPositive(server, 'test/static/file.png', done);
        });

        lab.test('Returns error if media type is not supported', function (done) {
            common.verifyNegative(server, 'test/static/file.gif', done);
        });

    });
});


test/common.js
'use strict';

const fs = require('fs');

const _ = require('lodash');
const Hapi = require('hapi');
const Code = require('code');

const FormData = require('form-data');
const streamToPromise = require('stream-to-promise');

const plugin = require('../');

const internals = {};

internals.append = function (file, fn) {

    const form = new FormData();
    form.append('text', 'Hello');
    form.append('file', fs.createReadStream(file));

    const options = {
        url: '/',
        method: 'POST',
        headers: form.getHeaders()
    };

    streamToPromise(form).then((payload) => {

        options.payload = payload;
        return fn(options);
    });
};


exports = module.exports = {};

exports.boostrap = function (config /*, options, fn */) {

    const options = _.isFunction(arguments[2]) ? arguments[1] : {};
    const fn = _.isFunction(arguments[2]) ? arguments[2] : arguments[1];

    const server = new Hapi.Server();
    const props = {
        register: plugin,
        options: options
    };

    server.connection();
    server.register(props, (err) => {

        if (err) {
            return fn(err);
        }

        server.route({
            path: '/',
            method: 'POST',
            config: {
                payload: config.payload,
                handler: config.handler
            }
        });

        return fn(null, server);

    });
};

exports.positive = function (server, file, done) {

    const tmp = 'test/tmp/file';

    const size = function (fileRef, fn) {

        fs.stat(fileRef, (err, stats) => {

            if (err) {
                return done(err);
            }

            return fn(null, stats.size);
        });
    };

    const verify = function (response) {

        Code.expect(response.statusCode).to.equal(200);

        size(file, (err, original) => {

            size(tmp, (err, copy) => {

                Code.expect(err).to.not.exist();
                Code.expect(copy).to.be.above(0);
                Code.expect(original).to.equal(copy);

                return done();
            });
        });
    };

    const inject = function (request) {

        server.inject(request, verify);
    };

    internals.append(file, inject);
};

exports.negative = function (server, file, done) {

    const tmp = 'test/tmp/file';

    const assert = function (err) {

        Code.expect(err).to.not.be.null();
        return done();
    };

    const verify = function (response) {

        Code.expect(response.statusCode).to.equal(415);
        fs.stat(tmp, assert);
    };

    const inject = function (request) {

        server.inject(request, verify);
    };

    internals.append(file, inject);
};


lib/index.js
'use strict';

const fs = require('fs');
const Readable = require('stream').Readable;

const _ = require('lodash');
const async = require('async');

const Subtext = require('subtext');
const Wreck = require('wreck');
const Boom = require('boom');

console.warn('This package is deprecated and not actively developed anymore. Read the deprecation notice (https://github.com/ruiquelhas/hapi-magic-filter/issues/16) for more details on how to move forward.')

const internals = {};

internals.defaults = {
    whitelist: {
        'bz2': '425a',
        'jpg': 'ffd8',
        'mp4': '3367',
        'ogg': '4f67',
        'pdf': '2550',
        'png': '8950',
        'tar': '7573'
    }
};

internals.mapWhitelist = function (original, fresh) {

    const filter = function (value, key) {

        return _.contains(fresh.allowed, key);
    };

    if (_.isArray(fresh.allowed)) {

        return _.pick(original, filter);

    }

    return fresh.allowed;
};

internals.isSupported = function (hex, wl, fn) {

    let whitelist;
    const ref = internals.defaults.whitelist;

    if (_.has(wl, 'allowed')) {
        whitelist = internals.mapWhitelist(ref, wl);
    }
    else {
        whitelist = ref;
    }

    const supported = _.pick(whitelist, (v) => _.isEqual(v, hex));

    if (_.isEmpty(supported)) {
        return fn(new Error());
    }

    return fn();
};

internals.validateBuffer = function (buffer, wl, fn) {

    const data = buffer.toString('hex').slice(0, 4);
    return internals.isSupported(data, wl, fn);
};

internals.validateStream = function (stream, wl, fn) {

    const done = function (err) {

        if (err) {
            return fn(err);
        }

        return fn();
    };

    const validate = function (data) {

        return internals.validateBuffer(data, wl, done);
    };

    stream.on('data', validate);
};

internals.validateTemporaryFile = function (file, wl, fn) {

    const stream = fs.createReadStream(file);
    return internals.validateStream(stream, wl, fn);
};

internals.validateFormFields = function (form, wl) {

    return function (key, fn) {

        const field = form[key];

        if (field instanceof Buffer) {
            return internals.validateBuffer(field, wl, fn);
        }

        if (_.has(field, 'path')) {
            const file = field.path;
            return internals.validateTemporaryFile(file, wl, fn);
        }

        if (field instanceof Readable) {
            const buffer = field._data;
            return internals.validateBuffer(buffer, wl, fn);
        }

        return fn();
    };
};

internals.parseBufferPayload = function (stream, wl, fn) {

    const _validate = function (err, parsed) {

        if (err) {
            return fn(err);
        }

        const payload = parsed.payload;
        const items = _.keys(payload);
        const iterator = internals.validateFormFields(payload, wl);

        async.each(items, iterator, fn);
    };

    const options = {
        output: 'data',
        parse: true
    };

    return Subtext.parse(stream, null, options, _validate);
};

internals.validateRawDataPayload = function (request, wl, fn) {

    const stream = Wreck.toReadableStream(request.payload);
    stream.headers = request.headers;

    return internals.parseBufferPayload(stream, wl, fn);
};

internals.validateRawFilePayload = function (request, wl, fn) {

    const _read = function (err, data) {

        if (err) {
            return fn(err);
        }

        const stream = Wreck.toReadableStream(data);
        stream.headers = request.headers;

        internals.parseBufferPayload(stream, wl, fn);
    };

    fs.readFile(request.payload.path, _read);
};

internals.validateRawStreamPayload = function (request, wl, fn) {

    const stream = request.payload;
    stream.headers = request.headers;

    return internals.parseBufferPayload(stream, wl, fn);
};

internals.validate = function (request, wl, fn) {

    const payload = request.payload;

    if (payload instanceof Buffer) {
        return internals.validateRawDataPayload(request, wl, fn);
    }

    if (_.has(payload, 'path')) {
        return internals.validateRawFilePayload(request, wl, fn);
    }

    if (payload instanceof Readable) {
        return internals.validateRawStreamPayload(request, wl, fn);
    }

    const items = _.keys(payload);
    const iterator = internals.validateFormFields(payload, wl);

    async.each(items, iterator, fn);
};

internals.attach = function (options) {

    const method = 'post';
    const mime = 'multipart/form-data';

    return function (request, reply) {

        const done = function (err) {

            if (err) {
                return reply(Boom.unsupportedMediaType());
            }

            return reply.continue();
        };

        if (request.method === method && request.mime === mime) {
            options || (options = []); //jshint ignore: line

            return internals.validate(request, options, done);
        }

        return reply.continue();
    };
};

exports.register = function (plugin, options, next) {

    const handler = internals.attach(options);
    plugin.ext('onPreHandler', handler);

    return next();
};

exports.register.attributes = {
    name: 'hapi-magic-filter',
    version: '1.1.0'
};
